### ISSUE TITLE
what is difference between adam with learning rate lr0 & lrf  & cos_lr ?

### ISSUE THREAD

**Issue by aiakash** [ultralytics/ultralytics#2026]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

```
lr0: 0.01  # initial learning rate (i.e. SGD=1E-2, Adam=1E-3)
lrf: 0.01  # final learning rate (lr0 * lrf)
```
i want to use adam 
so my lr is 0: 0.01 ?
& lrf:0.01 or also i need to change it to lrf to 0.001?

both should be changed or only lr0 i neeed to change?

2. what is the use of cos_lr? because lr0 & lrf already doinig the same work, so if i on cos_lr its taking cos_lr or adam lr values?

**Comment by glenn-jocher** [comment]
@aiakash if you are using Adam optimizer, you still need to change the initial learning rate `lr0` to `0.001`. The final learning rate `lrf` can remain the same as `0.01`. 

Regarding `cos_lr`, it is a technique known as cosine annealing, which lowers the learning rate smoothly instead of using a step function. Cosine annealing can help improve the model's accuracy by allowing it to find a more fine-grained minimum in the loss landscape. If you enable `cos_lr`, it will be used instead of `lrf` to lower the learning rate.

**Comment by aiakash** [comment]
@glenn-jocher 
lr0=0.001
lrf=0.01  (0.001*0.01)=0.00001  (so internally its taking this value)

1)im correct?
2)) lrf -(final learning rate) - name is saying final lr ,so when its applying this lr? on what bases its changing its learning rate?
2)may i know where is the code of lrf,  by default which lrf its using ? exponetial,linear,cosin etc

**Comment by glenn-jocher** [comment]
@aiakash 

1) Yes, your understanding is correct. If you set `lr0=0.001` and `lrf=0.01`, the final learning rate will be `0.00001`. 

2) The `lrf` is used as an annealing factor to reduce the learning rate over time. Its effect is visible after a certain number of training iterations, usually set by the `burn_in` value in the configuration file. The learning rate decreases linearly from `lr0` to `lrf` at a linear rate until the `burn_in` threshold is reached, then it continues to decrease at a slower rate using a predetermined schedule (e.g. step or cosine annealing). 

3) `lrf` (final learning rate) is a constant value used as a scaling factor to reduce the learning rate over time. It doesn't specify the decay schedule itself. By default, YOLOv8 uses step decay, which reduces the learning rate by a factor of 0.1 at certain epochs. However, you can enable cosine annealing by setting the `cosine_lr` parameter to `True`.
